### PR TITLE
Use better upper bound for scheduled_date_gmt when claiming actions.

### LINKED ISSUE #530: Diagnose intermittent deadlocks
Continuing discussion from #438 on intermittent deadlocks:

```
WordPress database error Deadlock found when trying to get lock; try restarting transaction for query UPDATE wp_actionscheduler_actions SET attempts = attempts+1, status='in-progress', last_attempt_gmt = '2020-04-20 17:00:02', last_attempt_local = '2020-04-20 18:00:02' WHERE action_id = 1877421 made by do_action('wp_ajax_as_async_request_queue_runner'), WP_Hook->do_action, WP_Hook->apply_filters, WP_Async_Request->maybe_handle, ActionScheduler_AsyncRequest_QueueRunner->handle, do_action('action_scheduler_run_queue'), WP_Hook->do_action, WP_Hook->apply_filters, ActionScheduler_QueueRunner->run, ActionScheduler_QueueRunner->do_batch, ActionScheduler_Abstract_QueueRunner->process_action, ActionScheduler_DBStore->log_execution, QM_DB->query
```
and
```
WordPress database error Deadlock found when trying to get lock; try restarting transaction for query UPDATE wp_actionscheduler_actions SET attempts = attempts+1, status='in-progress', last_attempt_gmt = '2020-04-20 17:02:08', last_attempt_local = '2020-04-20 18:02:08' WHERE action_id = 1877473 made by do_action_ref_array('action_scheduler_run_queue'), WP_Hook->do_action, WP_Hook->apply_filters, ActionScheduler_QueueRunner->run, ActionScheduler_QueueRunner->do_batch, ActionScheduler_Abstract_QueueRunner->process_action, ActionScheduler_DBStore->log_execution
```

### PR DESCRIPTION
**Pull request by peterfabian**
TL;DR is that the query that was in place was causing pretty bad congestion on the actions table due to the ordering/inefficient use of indexes. This PR is an attempt to optimize the action claiming query to minimize the deadlocks and lock timeouts. 

Hopefully closes #530 , together with #678.

##Testing instructions
1. create a new site
2. install & activate WC and action scheduler
3. generate at least 100k actions, or load the file with 1.3M actions into the database [wordpress-one-actions-only.sql.zip](https://github.com/woocommerce/action-scheduler/files/6206859/wordpress-one-actions-only.sql.zip)
(e.g. via phpmyadmin or via ssh `mysql -u root -p db_name < file.sql` ) and attach some function to action `create_next_action`)
4. enable multiple runners, e.g. by adding the code below
```
// mulitple queues
function eg_increase_action_scheduler_concurrent_batches( $concurrent_batches ) {
	return 10;
}
add_filter( 'action_scheduler_queue_runner_concurrent_batches', 'eg_increase_action_scheduler_concurrent_batches' );
```
5. start multiple (e.g. 5) runners from WP CLI, e.g. via `wp action-scheduler run --batch-size=100`
6. start `mysql` from ssh and run `SHOW PROCESSLIST` to see which queries are waiting for a long time for locks to get released. My output looked like this and the whole site was super slow:
```
MariaDB [(none)]> SHOW PROCESSLIST;
+-----+-------------+-----------+---------------+---------+------+--------------------------+------------------------------------------------------------------------------------------------------+----------+
| Id  | User        | Host      | db            | Command | Time | State                    | Info                                                                                                 | Progress |
+-----+-------------+-----------+---------------+---------+------+--------------------------+------------------------------------------------------------------------------------------------------+----------+
|   1 | system user |           | NULL          | Daemon  | NULL | InnoDB purge worker      | NULL                                                                                                 |    0.000 |
|   3 | system user |           | NULL          | Daemon  | NULL | InnoDB purge coordinator | NULL                                                                                                 |    0.000 |
|   2 | system user |           | NULL          | Daemon  | NULL | InnoDB purge worker      | NULL                                                                                                 |    0.000 |
|   4 | system user |           | NULL          | Daemon  | NULL | InnoDB purge worker      | NULL                                                                                                 |    0.000 |
|   5 | system user |           | NULL          | Daemon  | NULL | InnoDB shutdown handler  | NULL                                                                                                 |    0.000 |
| 138 | wp          | localhost | wordpress-one | Query   |   45 | Init for update          | UPDATE wp_actionscheduler_actions SET claim_id=60176, last_attempt_gmt='2021-03-23 15:08:03', last_a |    0.000 |
| 141 | root        | localhost | NULL          | Query   |    0 | Init                     | SHOW PROCESSLIST                                                                                     |    0.000 |
| 142 | wp          | localhost | wordpress-one | Query   |   19 | Init for update          | UPDATE wp_actionscheduler_actions SET claim_id=60177, last_attempt_gmt='2021-03-23 15:08:30', last_a |    0.000 |
| 143 | wp          | localhost | wordpress-one | Query   |   19 | Init for update          | UPDATE wp_actionscheduler_actions SET claim_id=60178, last_attempt_gmt='2021-03-23 15:08:30', last_a |    0.000 |
| 144 | wp          | localhost | wordpress-one | Query   |   16 | Sending data             | SELECT a.action_id FROM wp_actionscheduler_actions a WHERE 1=1 AND a.status='complete' AND a.last_at |    0.000 |
| 145 | wp          | localhost | wordpress-one | Query   |   11 | Sending data             | SELECT a.action_id FROM wp_actionscheduler_actions a WHERE 1=1 AND a.status='complete' AND a.last_at |    0.000 |
+-----+-------------+-----------+---------------+---------+------+--------------------------+------------------------------------------------------------------------------------------------------+----------+
11 rows in set (0.000 sec)
```
7. Now apply this branch and you shouldn't see any more stalled transactions (or UPDATE queries running for more than 1-2 secs)
```
MariaDB [wordpress-one]> SHOW PROCESSLIST;
+-----+-------------+-----------+---------------+---------+------+--------------------------+------------------------------------------------------------------------------------------------------+----------+
| Id  | User        | Host      | db            | Command | Time | State                    | Info                                                                                                 | Progress |
+-----+-------------+-----------+---------------+---------+------+--------------------------+------------------------------------------------------------------------------------------------------+----------+
|   1 | system user |           | NULL          | Daemon  | NULL | InnoDB purge worker      | NULL                                                                                                 |    0.000 |
|   2 | system user |           | NULL          | Daemon  | NULL | InnoDB purge worker      | NULL                                                                                                 |    0.000 |
|   3 | system user |           | NULL          | Daemon  | NULL | InnoDB purge worker      | NULL                                                                                                 |    0.000 |
|   4 | system user |           | NULL          | Daemon  | NULL | InnoDB purge coordinator | NULL                                                                                                 |    0.000 |
|   5 | system user |           | NULL          | Daemon  | NULL | InnoDB shutdown handler  | NULL                                                                                                 |    0.000 |
|  63 | root        | localhost | wordpress-one | Query   |    0 | Init                     | SHOW PROCESSLIST                                                                                     |    0.000 |
| 215 | root        | localhost | wordpress-one | Sleep   |  203 |                          | NULL                                                                                                 |    0.000 |
| 306 | wp          | localhost | wordpress-one | Query   |    0 | Init for update          | UPDATE wp_actionscheduler_actions SET claim_id=69077, last_attempt_gmt='2021-03-25 17:19:17', last_a |    0.000 |
| 307 | wp          | localhost | wordpress-one | Sleep   |    0 |                          | NULL                                                                                                 |    0.000 |
| 309 | wp          | localhost | wordpress-one | Query   |    0 | Sending data             | SELECT MAX(scheduled_date_gmt) FROM ( SELECT scheduled_date_gmt FROM wp_actionscheduler_actions a WH |    0.000 |
| 311 | wp          | localhost | wordpress-one | Query   |    0 | NULL                     | SELECT a.*, g.slug AS `group` FROM wp_actionscheduler_actions a LEFT JOIN wp_actionscheduler_groups  |    0.000 |
| 312 | wp          | localhost | wordpress-one | Query   |    0 | Sending data             | SELECT a.action_id FROM wp_actionscheduler_actions a WHERE 1=1 AND a.status='complete' AND a.last_at |    0.000 |
+-----+-------------+-----------+---------------+---------+------+--------------------------+------------------------------------------------------------------------------------------------------+----------+
12 rows in set (0.001 sec)
```

Note for future self: I generated the actions by putting this code into wc-core-functions:

```
add_action( 'shutdown', 'bootstrap' );

function bootstrap() {
	$pending_jobs = WC()->queue()->search(
		array(
			'per_page' => 1,
			'group'    => 'AS_PR_test_bootstrap',
		)
	);

	if ( ! $pending_jobs ) {
		$q = WC()->queue();
		$q->schedule_single( time() + 5, 'create_next_action', array( 0 ), 'AS_PR_test_bootstrap' );
	}
}

add_action( 'create_next_action', 'meh' );

function meh( $n ){
	$q = WC()->queue();
	for ( $i = 0; $i < 100; $i++ ){
		$q->schedule_single( time() + MINUTE_IN_SECONDS * $i, 'create_next_action', array( "$n.$i" ) );
	}
	sleep(1);
}
```

and then running the queue from WP CLI overnight
```
php8.0 wp-cli.phar action-scheduler run --batch-size=10
```

### CODE REVIEW

**Comment by peterfabian** [comment]
oops, this is a combined PR. closing and will open one with changes only relevant for claiming actions.
